lib/SILOptimizer/Transforms/PerformanceInliner.cpp

```diff
@@ -364,6 +366,37 @@ bool SILPerformanceInliner::isAutoDiffLinearMapWithControlFlow(
   return false;
 }
 
+// Checks if the function may be an AutoDiff VJP returning a partially
+// applied pullback. VJPs return a tuple of the original result and the
+// pullback, which may or may not be partially applied.
+bool SILPerformanceInliner::mayBeAutoDiffVJPReturningPartiallyAppliedPullback(
+    SILValue retVal) {
+  if (auto *ti = dyn_cast<TupleInst>(retVal)) {
+    if (ti->getNumOperands() == 2) {
+      auto val = ti->getOperand(1);
+
+      for (;;) {
+        if (auto *pai = dyn_cast<PartialApplyInst>(val)) {
```

Comment:
This seems to be quite a common pattern (also happening for pullbacks in `isAutoDiffLinearMapWithControlFlow`). Maybe we should have helper? It would be similar to now-obsolete LLVM's `stripPointerCasts` (before opaque pointers).

Reply:
There's parts which are similar. Let me see if I can reuse certain parts of the code.